Add cookbook details to message , support default webhook config from Slack UI, plus bug fixes

cknowles opened the pull request:
Work done for https://github.com/rackspace-cookbooks/chef-slack_handler/issues/17

Also resolves https://github.com/rackspace-cookbooks/chef-slack_handler/issues/15 and https://github.com/rackspace-cookbooks/chef-slack_handler/issues/16

Commit:
Add cookbook details to message

https://github.com/rackspace-cookbooks/chef-slack_handler/issues/17

Also resolves
https://github.com/rackspace-cookbooks/chef-slack_handler/issues/15
https://github.com/rackspace-cookbooks/chef-slack_handler/issues/16

Commit:
Merge remote-tracking branch 'rackspace-cookbooks/master'

# Conflicts:
#	files/default/slack_handler_webhook.rb

Commit:
Aligned message methods and add readme details

For https://github.com/rackspace-cookbooks/chef-slack_handler/issues/17

martinb3 commented:
This seems reasonable, but I'm wondering if this should be a major version bump, due to the changes to some of the default values. @c-knowles, do you agree?

cknowles commented:
Not sure it justifies a major bump as it will still work without the node.name as the default handle I believe. I wasn't able to check that end to end for the API key/slackr method as I can't find where to enable this. Do you know if it is deprecated with WebHooks taking its place? Every custom integration link I found seems to direct me back to WebHooks only.

From a pure backwards compatibility perspective then yes a major bump would be suitable. I can add to the pull request if that's what we want.

nicwaller commented:
I don't think a major version bump is warranted by these changes. SemVer permits breaking changes without a major version bump when you're still in 0.x development versions.

> 4. Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable. [SemVer.org - item 4](http://semver.org/#spec-item-4)

Of course, this raises the question: when should the 1.0.0 version be declared? In my opinion, a cookbook should declare 1.0.0 when it is first published to Supermarket.

> If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backwards compatibility, you should probably already be 1.0.0. [SemVer.org - when to release 1.0.0](http://semver.org/#how-do-i-know-when-to-release-100)

Given that this cookbook hasn't declared 1.0.0 yet, what would be normal in the context of the Chef community? I have observed that other cookbooks in the 0.x development stage will introduce breaking changes with a minor version bump _and include a note in the changelog_.

And what if this cookbook had already declared 1.0.0? Would these changes then warrant bumping the major version? The [chef-community CVP](https://chef-community.github.io/cvp/) would say that it warrants a major version bump because it "changes the value assigned to an attribute by the cookbook". But I don't quite agree with that. SemVer opines:

> Semantic Versioning is all about conveying meaning by how the version number changes. If these changes are important to your users, use the version number to inform them. [SemVer](http://semver.org/#what-if-i-inadvertently-alter-the-public-api-in-a-way-that-is-not-compliant-with-the-version-number-change-ie-the-code-incorrectly-introduces-a-major-breaking-change-in-a-patch-release)

It really comes down to this: how is the change important to users of this cookbook? I think most people would agree that this is providing new functionality (aha, the username and icon I chose when setting up my Slack integration are finally working!).

Commit:
Fixes for support for cookbook context

Changed `basic` to `off` for new cookbook context feature as it makes
more sense.
detail_level in slack_handler.rb was not taking the default correctly,
now is.
Default username for slackr integration now the node name again.
Defaults applicable to both handlers moved to default.rb.
Exposed cookbook detail level in slack_handler.rb.
https://github.com/rackspace-cookbooks/chef-slack_handler/issues/17

Commit:
Global attributes now work for webhooks

https://github.com/rackspace-cookbooks/chef-slack_handler/issues/18

Commit:
Improvements to README and CHANGELOG

Based on feedback from
https://github.com/rackspace-cookbooks/chef-slack_handler/pull/19,
added details to CHANGELOG and improved README.

Commit:
Bumped version number

Based on feedback in
https://github.com/rackspace-cookbooks/chef-slack_handler/pull/19,
reintroduced node name as default username for slackr integration and
bumped minor version.

cknowles commented:
Right, I've re-introduced the node name as default but only for the slackr integration. I've updated the CHANGELOG with all the changes plus clarified several parts of the README. As such, I've only bumped the minor version number as the default has only changed for the WebHook integration in order to add support for #16.

cknowles commented:
Would love to get this merged if possible. Do we need more tests? There's a bit of a lack of coverage in something like rspec.

shortdudey123 commented on files/default/slack_handler.rb:
```diff
@@ -57,21 +59,51 @@ def report
+    when "off"
+      return
+    when "all"
+      cookbooks = run_status.run_context.cookbook_collection
```
`run_status.run_context` appears to be `nil`

```
From: /var/chef/handlers/slack_handler_webhook.rb @ line 98 Chef::Handler::Slack#run_status_human_readable:

     96: def run_status_human_readable
     97:   require 'pry'
 =>  98:   binding.pry
     99:   run_status.success? ? "succeeded" : "failed"
    100: end

[1] pry(#<Chef::Handler::Slack>)> run_status.run_context
=> nil
[2] pry(#<Chef::Handler::Slack>)> 
```

cknowles replied:
I was going off [these chef docs](https://docs.chef.io/handlers.html#run-status-object) which do state some of the run_status properties are not always available. I've never seen this property as `nil` though. Please can you let me know how you ran it?  If we go with run_list as you say below then perhaps this goes away anyway.

martinb3 replied:
Can we make this a little more defensive, e.g.

```
if run_status && run_status.success?
  "success"
elsif run_status && !run_status.success?
  "failed"
else
  "unknown"
end
```

I think that would address the immediate issue of it sometimes being nil.

shortdudey123 replied:
`run_status` itself is not `nil`, its `run_status.run_context` that is `nil`

martinb3 replied:
Have you tested this during a chef converge? I was able to see these messages print out successfully when I tested.

shortdudey123 replied:
i have not tested what is currently in master now, i was doing tests based off of what was in master before this PR was merged.  I will update my stuff in a bit and make sure

shortdudey123 commented on files/default/slack_handler.rb:
```diff
@@ -57,21 +59,51 @@ def report
+      cookbooks = run_status.run_context.cookbook_collection
+      " using cookbooks #{cookbooks.values.map { |x| x.name.to_s + ' ' + x.version }}"
+    when "root"
+      root_cookbook = run_status.run_context.cookbook_collection.values.first
```
This does not really represent the "root" cookbook.  It only represents the one that is in the run list.  I think this should be the run_list instead

cknowles replied:
Please could you advise how we could access the `run_list`? Is it another global property? Just looking through the chef docs for it.

I found `run_status.node.run_list` has at least the root cookbook in it but didn't see the version used. Although in all the test runs setup for this project, the name seems to be the same as the value gained from this code. Happy to go with either, whatever is appropriate. Ideally if we swap, I'd need to find a way to include the version.

martinb3 replied:
I'm not really even clear on what a 'root cookbook' is. As I understand it, there's no concept in Chef -- so really this is just pulling the first of an arbitrary-order list.

shortdudey123 replied:
@martinb3 correct :) hence the runlist suggestion instead

shortdudey123 commented on files/default/slack_handler_webhook.rb:
```diff
@@ -96,4 +110,20 @@ def request_body(message, text_attachment)
+  def run_status_cookbook_detail(detail_level)
+    detail_level ||= @cookbook_detail_level
+    case detail_level
+    when "off"
```
Since this is the same as the default case, i am not sure that `when "off"` is needed

cknowles replied:
Updated in both usages - just removed the defaults complete for when `nil` needs to be returned.

shortdudey123 commented on files/default/slack_handler_webhook.rb:
```diff
@@ -82,12 +95,13 @@ def slack_message(message, webhook, text_attachment = nil)
       body[:icon_url] = @icon_url
     else
-      body[:icon_emoji] = @icon_emoji
+      body[:icon_emoji] = @icon_emoji unless @icon_emoji.nil?
```
Should the `else` be `elsif @icon_emoji` instead of putting an `unless` here?

cknowles replied:
Updated below.

shortdudey123 commented on CHANGELOG.md:
```diff
@@ -1,5 +1,15 @@
 # Changelog
 
+## v0.5.0 (2016-03-14)
```
Date has already passed :)

cknowles replied:
Updated below, removed the date for now and can be added again later when 0.5.0 is released.

Commit:
Fixes based on pull request feedback

For https://github.com/rackspace-cookbooks/chef-slack_handler/pull/19:
- Bug fix for slack_handler.rb slackr_options not referenced correctly
- Do not send icon_emoji of nil to slackr
- Tidy some of the if/else statements
- Removed date from 0.5.0 release in CHANGELOG, it can be added once
the release has been made

martinb3 commented:
Hey guys,

I took some of the suggestions from @shortdudey123 and the work already done by @c-knowles, and made a few tweaks (added better testing, removed the 'root cookbook' idea, etc)... please give master a try and see what you think.

Thanks!

cknowles commented:
Great, thanks!

The reason I was thinking about a "root" even though it's not a chef concept was because in many of our runs we run one parent cookbook which in turn tend to reference several others. In the kitchen tests for this project, the `all` setting for cookbook detail will produce `["slack_handler 0.4.0", "chef_handler 1.1.6"]`. In this case, it's good to see the version of the chef_handler used as it is not fixed in the source. In other cases where dependencies are fixed, I was thinking that many users would only interested in the first cookbook in the list for a basic Slack message since the dependencies are traceable from the parent source. I haven't tried on all of our projects yet but I think the `all` setting may cause the message to be particularly large in some cases (although it will still be useful).
